### PR TITLE
fix(client): fix trie_node_touched bug

### LINKED ISSUE #3020: Error in trie_node_touched
On betanet some nodes crash with
```
thread '<unnamed>' panicked at 'attempt to subtract with overflow', runtime/near-vm-logic/src/logic.rs:2014:47
stack backtrace:
   0:     0x55a1388c5cdd - backtrace::backtrace::libunwind::trace::h98d3444f0ad70429
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1:     0x55a1388c5cdd - backtrace::backtrace::trace_unsynchronized::h37f821d20053e6cd
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2:     0x55a1388c5cdd - std::sys_common::backtrace::_print_fmt::hff04dc7e915a3c85
                               at src/libstd/sys_common/backtrace.rs:78
   3:     0x55a1388c5cdd - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h61891c439a79f00d
                               at src/libstd/sys_common/backtrace.rs:59
   4:     0x55a13803ffcc - core::fmt::write::hdf236390fbd68d3d
                               at src/libcore/fmt/mod.rs:1069
   5:     0x55a1388c5474 - std::io::Write::write_fmt::h55f7fe3633ce766f
                               at src/libstd/io/mod.rs:1537
   6:     0x55a1388c4c95 - std::sys_common::backtrace::_print::h78339381acb79e1b
                               at src/libstd/sys_common/backtrace.rs:62
   7:     0x55a1388c4c95 - std::sys_common::backtrace::print::h525aa9f9405c80a1
                               at src/libstd/sys_common/backtrace.rs:49
   8:     0x55a1388c4c95 - std::panicking::default_hook::{{closure}}::hace90b8a218287e6
                               at src/libstd/panicking.rs:198
   9:     0x55a1388c44aa - std::panicking::default_hook::h160b6beccc49b335
                               at src/libstd/panicking.rs:218
  10:     0x55a1380a1004 - near_actix_utils::init_stop_on_panic::{{closure}}::{{closure}}::hacbef665bf176031
  11:     0x55a1388c3f3d - std::panicking::rust_panic_with_hook::h36f6804eea3f553a
                               at src/libstd/panicking.rs:481
  12:     0x55a1388c3de8 - rust_begin_unwind
                               at src/libstd/panicking.rs:385
  13:     0x55a13803db20 - core::panicking::panic_fmt::hd101a87121fa411f
                               at src/libcore/panicking.rs:89
  14:     0x55a13803d93c - core::panicking::panic::h02171c407fa1462f
                               at src/libcore/panicking.rs:52
  15:     0x55a138672cba - <FN as wasmer_runtime_core::typed_func::HostFunction<wasmer_runtime_core::typed_func::ExplicitVmCtx,(A,B,C),Rets>>::to_raw::wrap::h79f462137ddedd85
  16:     0x7fbcd6ea8b80 - <unknown>
thread '<unnamed>' panicked at 'Bad error case! Output is non-deterministic TypeId { t: 12504030849213737826 } "User supplied error: \"attempt to subtract with overflow\""', runtime/near-vm-runner/src/wasmer_runner.rs:165:21
stack backtrace:
   0:     0x55a1388c5cdd - backtrace::backtrace::libunwind::trace::h98d3444f0ad70429
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1:     0x55a1388c5cdd - backtrace::backtrace::trace_unsynchronized::h37f821d20053e6cd
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2:     0x55a1388c5cdd - std::sys_common::backtrace::_print_fmt::hff04dc7e915a3c85
                               at src/libstd/sys_common/backtrace.rs:78
   3:     0x55a1388c5cdd - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h61891c439a79f00d
                               at src/libstd/sys_common/backtrace.rs:59
   4:     0x55a13803ffcc - core::fmt::write::hdf236390fbd68d3d
                               at src/libcore/fmt/mod.rs:1069
   5:     0x55a1388c5474 - std::io::Write::write_fmt::h55f7fe3633ce766f
                               at src/libstd/io/mod.rs:1537
   6:     0x55a1388c4c95 - std::sys_common::backtrace::_print::h78339381acb79e1b
                               at src/libstd/sys_common/backtrace.rs:62
   7:     0x55a1388c4c95 - std::sys_common::backtrace::print::h525aa9f9405c80a1
                               at src/libstd/sys_common/backtrace.rs:49
   8:     0x55a1388c4c95 - std::panicking::default_hook::{{closure}}::hace90b8a218287e6
                               at src/libstd/panicking.rs:198
   9:     0x55a1388c44aa - std::panicking::default_hook::h160b6beccc49b335
                               at src/libstd/panicking.rs:218
  10:     0x55a1380a1004 - near_actix_utils::init_stop_on_panic::{{closure}}::{{closure}}::hacbef665bf176031
  11:     0x55a1388c3f3d - std::panicking::rust_panic_with_hook::h36f6804eea3f553a
                               at src/libstd/panicking.rs:481
  12:     0x55a1388c3de8 - rust_begin_unwind
                               at src/libstd/panicking.rs:385
  13:     0x55a1388c3d90 - std::panicking::begin_panic_fmt::hd7ab56e05bfb1154
                               at src/libstd/panicking.rs:339
  14:     0x55a138663c7b - near_vm_runner::wasmer_runner::<impl near_vm_runner::errors::IntoVMError for wasmer_runtime_core::error::RuntimeError>::into_vm_error::h4003a201681114de
  15:     0x55a13865d182 - near_vm_runner::runner::run_vm::h539dbe78311f26c5
  16:     0x55a13873e8cd - node_runtime::state_viewer::TrieViewer::call_function::h12dd92caf7619ca9
  17:     0x55a138712f7c - <neard::runtime::NightshadeRuntime as near_chain::types::RuntimeAdapter>::query::hc601a1a8ad80e3b5
  18:     0x55a13827845a - near_client::view_client::ViewClientActor::handle_query::hbb3a43cce30fe99a
  19:     0x55a1383dd6c1 - <actix::sync::SyncContextEnvelope<A,M> as actix::address::envelope::EnvelopeProxy>::handle::h0bac3534e8ff0e2e
  20:     0x55a13826a5f1 - std::sys_common::backtrace::__rust_begin_short_backtrace::h7c156e12049e1ec3
  21:     0x55a138268bb6 - core::ops::function::FnOnce::call_once{{vtable.shim}}::h877a99de2e50dd48
  22:     0x55a1388d7afa - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::hb4fe39e1689c3de4
                               at /rustc/a74d1862d4d87a56244958416fd05976c58ca1a8/src/liballoc/boxed.rs:1034
  23:     0x55a1388d7afa - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::hae1de892063d9208
                               at /rustc/a74d1862d4d87a56244958416fd05976c58ca1a8/src/liballoc/boxed.rs:1034
  24:     0x55a1388d7afa - std::sys::unix::thread::Thread::new::thread_start::h8b0ae441dc3343d4
                               at src/libstd/sys/unix/thread.rs:87
  25:     0x7fbcfabb66db - start_thread
  26:     0x7fbcfa6c788f - __clone
  27:                0x0 - <unknown>
```
when serving rpc.

### PR DESCRIPTION
**Pull request by mikhailOK**
Add read snapshots to storage and do refactoring where
Trie instances are no longer persisted.

Fixes #3020 

Test plan
---------
existing tests

### CODE REVIEW

File: core/store/src/db.rs [hunk 1]
```diff
@@ -305,27 +367,56 @@ impl Database for RocksDB {
         }
         Ok(self.db.write(batch)?)
     }
+
+    fn get_snapshot(self: Pin<Arc<Self>>) -> Box<dyn ReadSnapshot> {
```
Comment (ailisp) [inline core/store/src/db.rs:371]: Where's get_snapshot called. didn't see it the change

Comment (mikhailOK) [reply core/store/src/db.rs:371]: not called, just adding it. Will revert from this PR

File: core/store/src/trie/shard_tries.rs [hunk 2]
```diff
@@ -9,40 +9,41 @@ use near_primitives::types::{
...
 
-    pub fn get_trie_for_shard(&self, shard_id: ShardId) -> Arc<Trie> {
-        return Arc::clone(&self.tries[shard_id as usize]);
+    pub fn get_trie_for_shard(&self, shard_id: ShardId) -> Trie {
+        let store = Box::new(TrieCachingStorage::new(
+            self.store.clone(),
+            self.caches[shard_id as usize].clone(),
+            shard_id,
+        ));
+        return Trie::new(store, shard_id);
```
Comment (bowenwang1996) [inline core/store/src/trie/shard_tries.rs:38]:
```suggestion
        Trie::new(store, shard_id)
```

File: core/store/src/trie/trie_storage.rs [hunk 3]
```diff
@@ -130,44 +148,28 @@ impl TrieCachingStorage {
     /// # Errors
     /// StorageError::StorageInternalError if the storage fails internally.
     pub fn retrieve_rc(&self, hash: &CryptoHash) -> Result<u32, StorageError> {
-        let mut guard = self.cache.lock().expect(POISONED_LOCK_ERR);
-        if let Some(val) = guard.cache_get(hash) {
-            Self::vec_to_rc(val)
+        // Ignore cache to be safe. retrieve_rc is used only when writing storage and cache is shared with readers.
```
Comment (bowenwang1996) [inline core/store/src/trie/trie_storage.rs:151]: but don't we write to storage quite frequently (every block)?

**Comment by evgenykuzyakov** [comment]
Is it possible to reproduce the panic from the linked issue within a test?

File: core/store/src/trie/trie_storage.rs (hunk 3 above)
Comment (mikhailOK) [reply core/store/src/trie/trie_storage.rs:151]: most of the time retrieve_rc is called on new values, and we go to storage anyway

File: core/store/src/db.rs [hunk 4]
```diff
@@ -226,6 +227,7 @@ impl DBTransaction {
 pub struct RocksDB {
     db: DB,
     cfs: Vec<*const ColumnFamily>,
+    _pin: PhantomPinned,
```
Comment (matklad) [inline core/store/src/db.rs:230]: What's was the intention of pining related changes here (adding `PhantomPinned` and `Pin<dyn ...>`)? On the first glance, it seems as if they don't do anything useful here.

Comment (Longarithm) [reply core/store/src/db.rs:230]: cc @bowenwang1996 

I'm simply worried that we mention such change the second time in two weeks. First was https://github.com/near/nearcore/pull/5722.

Comment (bowenwang1996) [reply core/store/src/db.rs:230]: I am not sure 😅

**Pull request merged**